guides/micronaut-oracle-function-http/micronaut-oracle-function-http.adoc

```diff
@@ -0,0 +1,370 @@
+
+include::{commondir}/common-default-package.adoc[]
+
+include::{commondir}/common-annotationprocessors.adoc[]
+
+=== Dependencies
+
+Add a dependency for `oci-java-sdk-core` so we have access to classes for managing compute instances:
+
+dependency:oci-java-sdk-core[groupId=com.oracle.oci.sdk]
+
+=== InstanceData
+
+Then create an `InstanceData` DTO class to represent Compute Instance properties:
+
+source:InstanceData[]
+
+=== MicronautguideController
+
+The generated application contains a `MicronautguideController` class which is good for getting started, but we'll update it to demonstrate working with OCI SDK APIs, in this case working with Compute Instances.
+
+Replace the generated `MicronautguideController` with this:
+
+source:MicronautguideController[]
+
+<1> The controller's root uri  is `/compute`
+<2> Here we dependency-inject the SDK `ComputeClient` instance
+<3> This endpoint accepts `GET` requests and returns current instance properties for the specified instance OCID
+<4> This endpoint accepts `POST` requests and starts the specified instance if it is stopped
+<5> This endpoint accepts `POST` requests and stops the specified instance if it is running
+
+== Testing the Application
+
+NOTE: The code in this guide should work with recent versions of Micronaut 2.5+, but the tests require at least version 2.5.7
+
+We need to update the generated `MicronautguideControllerTest` to test the changes we made in `MicronautguideController`.
+
+This will be a unit test, so we'll need some mock beans to replace the beans auto-registered by the `oracle-cloud-sdk` module which make requests to Oracle Cloud. We also need a workaround for the strict classloader isolation used by https://fnproject.io/[Fn Project] (which powers Oracle Cloud Functions).
+
+When running our tests, the test and mock classes are loaded by the regular Micronaut classloader, but the function invocations are made in a custom `Fn` classloader, so it is not _directly_ possible to share state between the two, which complicates traditional approaches to mocking. There is support in `Fn` for sharing classes however, so instead of setting values in mock instances from the test to be used by the controller we'll set values in a `MockData` helper class that's registered as a class that's shared between classloaders.
+
+Create the `MockData` class:
+
+test:mock/MockData[]
+
+Next create the `MockAuthenticationDetailsProvider` class. The methods all return `null` since they won't be called; the bean merely needs to exist for dependency injection:
+
+test:mock/MockAuthenticationDetailsProvider[]
+
+<1> The `AuthCachingPolicy` annotation is used to disable caching; without this we would need to provide a valid private key since the provider methods would be invoked when constructing SDK client classes
+<2> We use `@Replaces` to replace a previously auto-registered `ConfigFileAuthenticationDetailsProvider` bean (which happens if $HOME/.oci/config exists)
```

Comment:
`ConfigFileAuthenticationDetailsProvider` is just one possible auth method, you could be using instance principals or resource principals so just rephrase to `... to replace the previously auto-configuration Oracle Cloud authentication method`